osquery/tables/system/windows/windows_eventlog.cpp

```diff
@@ -0,0 +1,271 @@
+        EvtClose(events[i]);
+        continue;
+      }
+
+      EvtClose(events[i]);
+
+      Row row;
+      std::wstringstream xml_event;
+      xml_event << renderedContent.data();
+      auto status = parseWelXml(context, xml_event.str(), row);
+      if (status.ok()) {
+        yield(TableRowHolder(new DynamicTableRow(std::move(row))));
+      }
+    }
+
+    ret = EvtNext(
+        queryResults, kNumEventsBlock, events.data(), INFINITE, 0, &numEvents);
+  }
+  if (ERROR_NO_MORE_ITEMS != GetLastError()) {
+    // No need to close the handler after error; The query
+    // EvtClose will also close all the event handler
+    VLOG(1) << "EvtNext failed with error " << GetLastError();
+  }
+}
+
+void genXfilterFromConstraints(QueryContext& context, std::string& xfilter) {
+  std::vector<std::string> xfilterList;
+
+  auto eids = context.constraints["eventid"].getAll(EQUALS);
+  if (!eids.empty()) {
+    xfilterList.emplace_back(
+        "(EventID=" + osquery::join(eids, ") or (EventID=") + ")");
+  }
+
+  auto pids = context.constraints["pid"].getAll(EQUALS);
+  if (!pids.empty()) {
+    xfilterList.emplace_back(
+        "(Execution[@ProcessID=" +
+        osquery::join(pids, "]) or (Execution[@ProcessID=") + "])");
+  }
+
+  auto times = context.constraints["time_range"].getAll(EQUALS);
+  auto timestamps = context.constraints["timestamp"].getAll(EQUALS);
+  if (!times.empty()) {
+    auto datetime = *times.begin();
+    auto time_vec = osquery::split(datetime, ";");
+
+    if (time_vec.size() == 1) {
+      auto _start = time_vec.front();
+      xfilterList.emplace_back("TimeCreated[@SystemTime&gt;='" + _start + "']");
+    } else if (time_vec.size() == 2) {
+      auto _start = time_vec.front();
+      auto _end = time_vec.at(1);
+      xfilterList.emplace_back("TimeCreated[@SystemTime&gt;='" + _start +
+                               "' and @SystemTime&lt;='" + _end + "']");
+    }
+  } else if (!timestamps.empty()) {
+    auto time_diff = *timestamps.begin();
+    xfilterList.emplace_back(
+        "TimeCreated[timediff(@SystemTime) &lt;= " + time_diff + "]");
+  }
+
+  xfilter = xfilterList.empty()
+                ? "*"
+                : "*[System[" + osquery::join(xfilterList, " and ") + "]]";
+}
+
+bool shouldHandleXpath(QueryContext& context) {
+  // xpaths are mutually execlusive and can't be used with itself
+  // and the other constraints
+  auto xpaths = context.constraints["xpath"].getAll(EQUALS);
+  if (xpaths.size() > 1) {
+    return false;
+  }
+
+  return !(context.hasConstraint("channel", EQUALS) ||
+           context.hasConstraint("time_range", EQUALS) ||
+           context.hasConstraint("timestamp", EQUALS));
+}
+
+void genWindowsEventLog(RowYield& yield, QueryContext& context) {
+  std::set<std::pair<std::string, std::string>> xpath_set;
+  auto hasXpath = context.hasConstraint("xpath", EQUALS);
+
+  if (hasXpath && !shouldHandleXpath(context)) {
+    LOG(WARNING) << "Error : xpaths are mutually exclusive and"
```

Comment:
If this is a `WARNING` do not include "Error".

I recommend `"Xpaths are mutually exclusive and cannot be used with constraints (channel, time_range, or timestamp)"`